### PR TITLE
Update documentation of the extension

### LINKED ISSUE #22: Document the JS
We should generate API docs for our JS code. We could try this with [autoapi](https://sphinx-autoapi.readthedocs.io/) and if that doesn't work, we should try [sphinx-js](https://sphinx-js-howto.readthedocs.io/en/latest/)

### PR DESCRIPTION
**Pull request by dojutsu-user**
Fixes #22

### CODE REVIEW

File: docs/js-api-reference.rst [hunk 1]
```diff
@@ -0,0 +1,29 @@
+JavaScript API Reference
+========================
+
+Following are the functions that are defined in `rtd_sphinx_search.js`_,
+You can overwrite any of them, if any kind of customizations are required.
```
Comment (ericholscher) [inline docs/js-api-reference.rst]: How can you overwrite them? Would be useful to have a quick example perhaps, if we want folks to do it.

**Comment by dojutsu-user** [comment]
@ericholscher Updated the docs.

File: docs/_ext/hidden_code_block.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+"""Simple, inelegant Sphinx extension which adds a directive for a
```
Comment (ericholscher) [inline docs/_ext/hidden_code_block.py]: Why did we add this to our repo instead of as a dependency to the requirements? Also it doesn't credit the URL where we actually got it from..

File: docs/js-api-reference.rst [hunk 3]
```diff
@@ -25,5 +25,33 @@ You can overwrite any of them, if any kind of customizations are required.
...
+
+    /**
+     * Return the required search bar.
+     */
+    function getInputField() {
+        var custom_search_bar = document.querySelector('.my-custom-search-bar');
+        return custom_search_bar;
+    }
+
+And then, you can `include this custom.js file in your docs`_.
```
Comment (ericholscher) [inline docs/js-api-reference.rst]: Does it have to be included after our file? I don't quite understand how this is working -- are we not namespacing our functions at all?

File: docs/_ext/hidden_code_block.py (hunk 2 above)
Comment (dojutsu-user) [reply docs/_ext/hidden_code_block.py]: It is not uploaded to pypi.
The instructions for installation say to do it this way  -- http://scopatz.github.io/hiddencode/#installation

File: docs/js-api-reference.rst (hunk 3 above)
Comment (dojutsu-user) [reply docs/js-api-reference.rst]: I tested this approach and there was this problem:
- The custom js file is included before rtd_sphinx_search.js

Then to overwrite a function, I found a way:
In our `rtd_sphinx_search.js`, wrap every function like this:

```js
if (!window.hasOwnProperty("function_name")) {
    window.function_name = () => {
        // function definition
    }
```

This approach works, I tested this locally.

File: docs/custom-design.rst [hunk 4]
```diff
@@ -9,7 +9,107 @@ is given below for the reference:
 
 .. code-block:: html
 
-    <!-- Coming Soon -->
+    <div class="search__outer__wrapper search__backdrop">
```
[the remark this answers is not in the document]
Comment (dojutsu-user) [reply docs/custom-design.rst]: This can be just a reference if anyone don't want to lookup with the chrome inspect element and finding the right css class to overwrite.

File: docs/js-api-reference.rst (hunk 3 above)
Comment (dojutsu-user) [reply docs/js-api-reference.rst]: I haven't done namespacing...
Should I add `RTD` before every variable?

File: docs/_ext/hidden_code_block.py (hunk 2 above)
Comment (ericholscher) [reply docs/_ext/hidden_code_block.py]: We should at least mention the URL where we got it from.

File: docs/js-api-reference.rst (hunk 3 above)
Comment (ericholscher) [reply docs/js-api-reference.rst]: Almost certainly we should be namespacing this code. I don't know the best way to do this in JS, @davidfischer? I think probably with some kind of object that contains all the functions.

File: docs/js-api-reference.rst [hunk 5]
```diff
@@ -25,5 +25,33 @@ You can overwrite any of them, if any kind of customizations are required.
 .. js:autofunction:: showSearchModal
 .. js:autofunction:: removeSearchModal
 
+Example: Overwriting A JavaScript Function
```
Comment (davidfischer) [inline docs/js-api-reference.rst]:
```suggestion
Example: Overriding A JavaScript Function
```

File: docs/js-api-reference.rst (hunk 3 above)
Comment (davidfischer) [reply docs/js-api-reference.rst]: We definitely want to namespace. I would go with `READTHEDOCS` so functions would be `READTHEDOCS. getInputField()`.

You can check whether `READTHEDOCS` exists with `typeof READTHEDOCS === 'undefined'`. Generally the order of JavaScript being included can be controlled but even then you can always wait until DOM ready to actually override anything.

The major reason for this is to not pollute the global namespace and to limit all our methods and data to one place. I would also like to eventually put `READTHEDOCS_DATA` into there (eg. `READTHEDOCS.BUILD_DATA`).

File: docs/js-api-reference.rst (hunk 1 above)
Comment (davidfischer) [reply docs/js-api-reference.rst]: Shouldn't this be "override".

File: docs/js-api-reference.rst [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+.. js:autofunction:: removeResults
+.. js:autofunction:: getErrorDiv
+.. js:autofunction:: fetchAndGenerateResults
+.. js:autofunction:: generateAndReturnInitialHtml
+.. js:autofunction:: showSearchModal
+.. js:autofunction:: removeSearchModal
+
+Example: Overwriting A JavaScript Function
+------------------------------------------
+
```
Comment (ericholscher) [inline docs/js-api-reference.rst]: Let's remove this section for now, ship this PR, and then do namespacing in another PR.
